src/test/pthreads/test_pthreads.rs

```diff
@@ -0,0 +1,537 @@
+    }
+}
+
+struct MuxTry {
+    mux1: libc::pthread_mutex_t,
+    mux2: libc::pthread_mutex_t,
+    cond: libc::pthread_cond_t,
+    pub num_locked: u32,
+    pub num_not_locked: u32,
+}
+
+impl MuxTry {
+    /// Create a new MuxTry with allocated pthread_mux_t and pthread_cond_t vars
+    ///
+    /// Returns error message if allocation fails
+    pub fn new() -> Result<Self, String> {
+        let mut rv = Self {
+            mux1: unsafe { std::mem::zeroed() },
+            mux2: unsafe { std::mem::zeroed() },
+            cond: unsafe { std::mem::zeroed() },
+            num_locked: 0,
+            num_not_locked: 0,
+        };
+        mutex_init(rv.mux1_ptr())?;
+        mutex_init(rv.mux2_ptr())?;
+        cond_init(rv.cond_ptr())?;
+        Ok(rv)
+    }
+
+    /// Returns a mutable pointer to first inner pthread_mutex_t
+    #[inline(always)]
+    pub fn mux1_ptr(&mut self) -> *mut libc::pthread_mutex_t {
+        &mut self.mux1 as *mut _
+    }
+
+    /// Returns a mutable pointer to second inner pthread_mutex_t
+    #[inline(always)]
+    pub fn mux2_ptr(&mut self) -> *mut libc::pthread_mutex_t {
+        &mut self.mux2 as *mut _
+    }
+
+    /// Returns a mutable pointer to second inner pthread_mutex_t
+    #[inline(always)]
+    pub fn cond_ptr(&mut self) -> *mut libc::pthread_cond_t {
+        &mut self.cond as *mut _
+    }
+}
+
+impl Drop for MuxTry {
+    fn drop(&mut self) {
+        mutex_destroy(self.mux1_ptr());
+        mutex_destroy(self.mux2_ptr());
+        cond_destroy(self.cond_ptr());
+    }
+}
+
+pub struct Attr {
+    attr: libc::pthread_attr_t,
+}
+
+impl Attr {
+    /// Create a new Attr with allocated pthread_attr_t variable
+    ///
+    /// Returns error message if allocation fails
+    pub fn new() -> Result<Self, String> {
+        let mut rv = Self {
+            attr: unsafe { std::mem::zeroed() },
+        };
+        attr_init(rv.ptr())?;
+        Ok(rv)
+    }
+
+    /// Returns a mutable pointer to the inner pthread_attr_t
+    #[inline(always)]
+    pub fn ptr(&mut self) -> *mut libc::pthread_attr_t {
+        &mut self.attr as *mut _
+    }
+}
+
+impl Drop for Attr {
+    fn drop(&mut self) {
+        attr_destroy(self.ptr());
+    }
+}
+
+// Initialize an allocated pthread_attr_t
+//
+// Returns error if pointer is null or pthread_attr_init fails
+//
+// Undefined behavior if called on an already initialized pthread_attr_t,
+// without first calling attr_destroy
+#[inline(always)]
+fn attr_init(ptr: *mut libc::pthread_attr_t) -> Result<(), String> {
+    if ptr == std::ptr::null_mut() {
+        Err("null pthread_mutex_t pointer".into())
+    } else if unsafe { libc::pthread_attr_init(ptr) } < 0 {
+        Err("pthread_attr_init failed".into())
+    } else {
+        Ok(())
+    }
+}
+
+// Destroy a pthread_attr_t
+#[inline(always)]
+fn attr_destroy(ptr: *mut libc::pthread_attr_t) {
+    if ptr != std::ptr::null_mut() {
+        unsafe { libc::pthread_attr_destroy(ptr) };
+    }
+}
```

Comment:
For any of the functions that do more than just a cast and where an error could occur, I don't think we want any of the `#[inline(always)]` in the test code. Performance isn't an issue here, and when things go wrong, it's useful to have a full backtrace.